File: 프로그래머스/lv2/12909. 올바른 괄호/올바른 괄호.java
import java.util.*;

class Solution {
    boolean solution(String s) {
        boolean answer = true;
        
        int count=0;

        // 스택 문제인데 스택으로 풀면 시간초과가 나는 문제가 있다?
        for(int i=0; i<s.length(); i++){
            if(s.charAt(i) == '('){     // 여는 괄호는 +1
                count++;
            }
            else if(s.charAt(i) == ')'){                       // 닫는 괄호는 -1
                count--;
            }
            if(count < 0){      // "())"같은 입력이 들어오면 false
                return false;
            }
        }
        
        if(count == 0){     // 괄호를 더한 합이 0 -> 짝이 맞는 경우
            answer = true;
        }
        else{
            answer = false;
        }
        
        return answer;
    }
}

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.py
def solution(s):
    answer = True
    
    sum_ = 0
    
    for i in range(len(s)):
        if s[i] == '(':
            sum_ += 1
        else:
            sum_ -= 1
        
        if sum_ < 0:
            return False
        
    if sum_ == 0:
        answer = True
    else:
        answer = False

    return answer